vocode/streaming/agent/groq_agent.py

```diff
@@ -0,0 +1,211 @@
+import os
+import random
+from typing import Any, AsyncGenerator, Dict, List, Optional, TypeVar, Union
+
+import sentry_sdk
+from loguru import logger
+from groq import AsyncGroq
+
+from vocode import sentry_span_tags
+from vocode.streaming.action.abstract_factory import AbstractActionFactory
+from vocode.streaming.action.default_factory import DefaultActionFactory
+from vocode.streaming.agent.base_agent import GeneratedResponse, RespondAgent, StreamedResponse
+from vocode.streaming.agent.groq_utils import (
+    format_groq_chat_messages_from_transcript,
+    vector_db_result_to_groq_chat_message,
+    groq_get_tokens
+)
+from vocode.streaming.agent.streaming_utils import collate_response_async, stream_response_async
+from vocode.streaming.models.actions import FunctionCallActionTrigger
+from vocode.streaming.models.agent import GroqAgentConfig
+from vocode.streaming.models.events import Sender
+from vocode.streaming.models.message import BaseMessage, BotBackchannel, LLMToken
+from vocode.streaming.models.transcript import Message
+from vocode.streaming.vector_db.factory import VectorDBFactory
+from vocode.utils.sentry_utils import CustomSentrySpans, sentry_create_span
+
+
+class GroqAgent(RespondAgent[GroqAgentConfig]):
+    groq_client: AsyncGroq
+
+    def __init__(
+        self,
+        agent_config: GroqAgentConfig,
+        action_factory: AbstractActionFactory = DefaultActionFactory(),
+        vector_db_factory=VectorDBFactory(),
+        **kwargs,
+    ):
+        super().__init__(
+            agent_config=agent_config,
+            action_factory=action_factory,
+            **kwargs,
+        )
+        self.groq_client = AsyncGroq(api_key=os.environ.get("GROQ_API_KEY"))
+
+        if not self.groq_client.api_key:
+            raise ValueError("GROQ_API_KEY must be set in environment or passed in")
+
+        if self.agent_config.vector_db_config:
+            self.vector_db = vector_db_factory.create_vector_db(self.agent_config.vector_db_config)
+
+    def get_functions(self):
+        assert self.agent_config.actions
+        if not self.action_factory:
+            return None
+        return [
+            self.action_factory.create_action(action_config).get_openai_function()
+            for action_config in self.agent_config.actions
+            if isinstance(action_config.action_trigger, FunctionCallActionTrigger)
+        ]
```

Comment:
Just to confirm, were you able to test function calling after implementing this? I took a quick look at the [Groq Docs](https://console.groq.com/docs/tool-use#tools-specifications) and it looks like they mention using `tools` whereas we still use `functions`.

Reply:
If the function calling doesn't work, let's disable it for now. So you can delete `get_functions()` and lines 76-77.

Reply:
Function calling is working with the code as it is atm, it was tested. Wanna keep it?

Reply:
Awesome yeah let's keep it!